### PR TITLE
[ios][tests][clipboard] Try writing expo-module iOS unit tests

### PR DESCRIPTION
**Pull request by barthap**
# Why

**This PR is an experiment on writing iOS unit tests for expo-module packages. Its purpose is to examine the current possibilities of testing expo-modules that use Sweet API and discover any related issues/difficulties. Knowing these should help develop the `expo-modules-test-core` framework in the future. This PR may serve as a playground for its development.**

I started with `expo-clipboard` as it seemed pretty simple (it turned out not 😅) to test and I know it well.

Resolves ENG-3326, part of ENG-3888

# How

> Feel free to discuss any of these in the comments 😉 

List of things done and problems encountered:
- Added test subspec to `expo-clipboard` and listed it in iOS client Podfile
- Created `ClipboardModuleSpec` class - a test suite for the `expo-clipboard` module
- Currently, created a test `ModuleHolder` instance and a `testModuleFunction()` helper to call module functions. Many possibilities to improve here:
  - [ ] Move this logic to e-m-c / e-m-test-core - it shouldn't be defined per-module
  - [ ] Support both sync and async functions
  - [ ] Don't enforce nullable result type as closure argument. Maybe do something that looks fully synchronous - get rid of the `waitUntil` closure.
  - [ ] Support for providing already-converted function args (enums, record instances etc.) instead of the primitive dicts/strings. 
  - [ ] Support optional/default args (either as `nil` or not provided at all) - especially records with defaults
- Had big issue with `UIPasteboard` returning `null` contents:
  ```swift
  it("copies and pastes a string") {
    UIPasteboard.general.string = "expo"
    expect(UIPasteboard.general.hasStrings) == true // ok
    expect(UIPasteboard.general.string) == "expo"     // expected expo, got nil
  }
  ```
  It happens because UIPasteboard doesn't return its contents when the app is in the background. I couldn't get over it - haven't found any possibility of making unit tests being run as a foreground app. It would be possible in UI tests though.
  To overcome that, I created my own mock `UIPasteboard` class and swizzled the getter of `UIPasteboard.general` to return my mock instance.
  > From @tsapeta: _"I wouldn't use unit tests for anything that comes from UIKit. I'm not expecting that every module will have unit tests."_

  However, I personally don't feel bad for mocking the `UIPasteboard` in this case.

# Test Plan

All new clipboard tests should pass

### CODE REVIEW

File: packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift [hunk 1]
```diff
@@ -0,0 +1,353 @@
+import ExpoModulesTestCore
+
+@testable import ExpoModulesCore
+@testable import ExpoClipboard
+import UIKit
+import MobileCoreServices
+
+class MockUIPasteboard: UIPasteboard {
```
Comment (tsapeta) [inline packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift:8]: Could you move this to the separate file? This class seems to be quite big 😉 
Also please, prettify this code 😛

File: packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  var _items: [String : Any] = [:]
+  
+  override var items: [[String : Any]]
+  {
+    get { return [_items] }
+    set(newItems) { _items = newItems.count > 0 ? newItems[0] : [:] }
+  }
```
Comment (tsapeta) [inline packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift:15]:
```suggestion
  override var items: [[String: Any]] {
    get {
      return [_items]
    }
    set(newItems) {
      _items = newItems.count > 0 ? newItems[0] : [:]
    }
  }
```

File: packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+extension UIPasteboard {
+  struct StaticVars {
+    static var mockPastebaord = MockUIPasteboard()
+  }
+  
+  @objc dynamic class var swizzledGeneralPasteboard: UIPasteboard
+  { get { return StaticVars.mockPastebaord } }
```
Comment (tsapeta) [inline packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift:69]:
```suggestion
  @objc
  dynamic class var swizzledGeneralPasteboard: UIPasteboard {
    get {
      return StaticVars.mockPastebaord
    }
  }
```

File: packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        UIPasteboard.general.image = UIImage()
+        
+        testModuleFunction("hasStringAsync", args: []) { (result: Bool?) in
+          expect(result) == false
+        }
+      }
+    }
+    
+    
+
```
Comment (tsapeta) [inline packages/expo-clipboard/ios/Tests/ClipboardModuleSpec.swift:204]:
```suggestion

```

**Review by bbarthec — APPROVED** [review]
I'm ok with this PR 😉

**Review by tsapeta — APPROVED** [review]

**Pull request merged**